### PR TITLE
Bugfix: #2948.  Persist Devices preference changes.

### PR DESCRIPTION
**Pull request by ksoze95**
Resolves: #2948 

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/prefs/DevicePrefs.cpp [hunk 1]
```diff
@@ -410,6 +412,11 @@ bool DevicePrefs::Commit()
       else
          AudioIORecordingSource.Reset();
       AudioIORecordChannels.Write(mChannels->GetSelection() + 1);
+
+      AudioIORecordingDevice.Commit();
+      AudioIORecordingSourceIndex.Commit();
+      AudioIORecordingSource.Commit();
+      AudioIORecordChannels.Commit();
```
Comment (Paul-Licameli) [inline src/prefs/DevicePrefs.cpp:419]: I am not willing to accept this very special case fix of what might be an error of commit d9a873a that introduced SettingsTransaction.  It was not intended that individual settings variables should require commits like this.

Do you instead have an explanation why the line `mTransaction->Commit();` in PrefsDialog.cpp fails to have the effect it should?

Can you investigate, verify that d9a873a is the first bad commit in 3.2 showing this bug?

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]

**Comment by Paul-Licameli** [comment]
Kit:  I was bothered enough that I debugged this a bit myself.

`AudioIOPlaybackDevice` is written in `DevicePrefs::Commit` and the later `mTransaction->Commit()` in `PrefsDialog::OnOk` should persist the setting but doesn't -- why not?

Because in between, there is re-reading of `AudioIOPlaybackDevice` that destroys the un-committed change, and that's a more general bug in Prefs.h that should be fixed.

See that there are two overloads of `ReadWithDefault`.  One of them is correct in case the settings object already has a value, but the other is incorrect.

**Review by Paul-Licameli — APPROVED** [review]
Very simple!

**Comment by petersampsonaudacity** [comment]
Testing on W10 with @ksoze95 's latest branch build:  audacity-win-3.2.0-alpha-20220522+2d8ee49-x64-msvc2022 

This is **_not_** working properly:


See  Devices Prefs ignores user settings for changes of I/O devices #2948  for my test results

**Comment by Paul-Licameli** [comment]
I'm closing this PR and writing a different fix that addresses more fundamental errors in the Preferences system and doesn't contain special case code to write one of the settings.

**Comment by Paul-Licameli** [comment]
I'm closing this PR and writing a different fix that addresses more fundamental errors in the Preferences system and doesn't contain special case code to write one of the settings.

**Pull request closed without merging**